File: compiler/extensions/cpp/runtime/src/zserio/BitBuffer.cpp
#include <utility>
#include <cstring>

#include "zserio/CppRuntimeException.h"
#include "zserio/StringConvertUtil.h"
#include "zserio/HashCodeUtil.h"
#include "zserio/BitBuffer.h"

namespace zserio
{

BitBuffer::BitBuffer() : m_bitSize(0)
{
}

BitBuffer::BitBuffer(size_t bitSize) : m_buffer((bitSize + 7) / 8), m_bitSize(bitSize)
{
}

BitBuffer::BitBuffer(const std::vector<uint8_t>& buffer) :
        m_buffer(buffer), m_bitSize(8 * buffer.size())
{
}

BitBuffer::BitBuffer(const std::vector<uint8_t>& buffer, size_t bitSize) :
        m_bitSize(bitSize)
{
    const size_t byteSize = (bitSize + 7) / 8;
    if (buffer.size() < byteSize)
        throw CppRuntimeException("BitBuffer: Bit size " + convertToString(bitSize) +
                " out of range for given vector byte size " + convertToString(buffer.size()) + "!");

    m_buffer.assign(buffer.data(), buffer.data() + byteSize);
}

BitBuffer::BitBuffer(std::vector<uint8_t>&& buffer) :
        m_buffer(std::move(buffer)), m_bitSize(8 * m_buffer.size())
{
}

BitBuffer::BitBuffer(std::vector<uint8_t>&& buffer, size_t bitSize) :
        m_buffer(std::move(buffer)), m_bitSize(bitSize)
{
    const size_t byteSize = (bitSize + 7) / 8;
    if (m_buffer.size() < byteSize)
        throw CppRuntimeException("BitBuffer: Bit size " + convertToString(bitSize) +
                " out of range for given vector byte size " + convertToString(buffer.size()) + "!");
}

BitBuffer::BitBuffer(const uint8_t* buffer, size_t bitSize) :
        m_buffer(buffer, buffer + (bitSize + 7) / 8), m_bitSize(bitSize)
{
}

bool BitBuffer::operator==(const BitBuffer& other) const
{
    if (this != &other)
    {
        if (m_bitSize != other.m_bitSize)
            return false;

        const size_t byteSize = getByteSize();
        if (byteSize > 0)
        {
            if (byteSize > 1)
            {
                if (memcmp(getBuffer(), other.getBuffer(), byteSize - 1) != 0)
                    return false;
            }

            if (getMaskedLastByte() != other.getMaskedLastByte())
                return false;
        }
    }

    return true;
}

int BitBuffer::hashCode() const
{
    int result = ::zserio::HASH_SEED;
    const size_t byteSize = getByteSize();
    if (byteSize > 0)
    {
        if (byteSize > 1)
        {
            const uint8_t* lastElement = getBuffer() + byteSize - 1;
            for (const uint8_t* p = getBuffer(); p < lastElement; p++)
                result = calcHashCode(result, *p);
        }
        result = ::zserio::calcHashCode(result, getMaskedLastByte());
    }

    return result;
}

const uint8_t* BitBuffer::getBuffer() const
{
    return m_buffer.data();
}

uint8_t* BitBuffer::getBuffer()
{
    return m_buffer.data();
}

size_t BitBuffer::getBitSize() const
{
    return m_bitSize;
}

size_t BitBuffer::getByteSize() const
{
    return (m_bitSize + 7) / 8;
}

uint8_t BitBuffer::getMaskedLastByte() const
{
    const size_t roundedByteSize = m_bitSize / 8;
    const uint8_t lastByteBits = static_cast<uint8_t>(m_bitSize - 8 * roundedByteSize);

    return (lastByteBits == 0) ? m_buffer[roundedByteSize - 1] :
            (m_buffer[roundedByteSize] & (0xFF << (8 - lastByteBits)));
}

} // namespace zserio

File: compiler/extensions/python/runtime/src/zserio/bitbuffer.py
"""
The module implements abstraction for holding bit sequence.
"""

import typing

from zserio.exception import PythonRuntimeException
from zserio.hashcode import HASH_SEED
from zserio.hashcode import calc_hashcode_int32
from zserio.bitposition import bitsize_to_bytesize

class BitBuffer:
    """
    Bit buffer.

    Because bit buffer size does not have to be byte aligned (divisible by 8), it's possible that not all bits
    of the last byte are used. In this case, only most significant bits of the corresponded size are used.
    """

    def __init__(self, buffer: bytes, bitsize: typing.Optional[int] = None) -> None:
        """
        Constructs bit buffer from bytes buffer and bit size.

        :param buffer: Bytes-like buffer to construct from.
        :param bitsize: Number of bits stored in buffer to use.
        :raises PythonRuntimeException: If bitsize is out of range.
        """

        if bitsize is None:
            bitsize = len(buffer) * 8
        elif len(buffer) * 8 < bitsize:
            raise PythonRuntimeException(f"BitBuffer: Bit size '{bitsize}' out of range "
                                         f"for the given buffer byte size '{len(buffer)}'!")
        self._buffer: bytes = buffer
        self._bitsize: int = bitsize

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, BitBuffer):
            return False

        if self._bitsize != other._bitsize:
            return False

        bytesize = bitsize_to_bytesize(self._bitsize)
        if bytesize > 0:
            if bytesize > 1:
                if self._buffer[0:bytesize - 1] != other._buffer[0:bytesize - 1]:
                    return False

            if self._masked_last_byte() != other._masked_last_byte():
                return False

        return True

    def __hash__(self) -> int:
        result = HASH_SEED
        bytesize = bitsize_to_bytesize(self._bitsize)
        if bytesize > 0:
            if bytesize > 1:
                for element in self._buffer[0:bytesize - 1]:
                    result = calc_hashcode_int32(result, element)

            result = calc_hashcode_int32(result, self._masked_last_byte())

        return result

    @property
    def buffer(self) -> bytes:
        """
        Gets the underlying byte buffer.

        Not all bits of the last byte must be used.

        :returns: The underlying byte buffer.
        """
        return self._buffer

    @property
    def bitsize(self) -> int:
        """
        Gets the number of bits stored in the bit buffer.

        :returns: Size of the bit buffer in bits.
        """
        return self._bitsize

    def _masked_last_byte(self) -> int:
        rounded_bytesize = self._bitsize // 8
        last_byte_bits = self._bitsize - 8 * rounded_bytesize

        return (self._buffer[rounded_bytesize - 1] if last_byte_bits == 0 else
                self._buffer[rounded_bytesize] & (0xFF << (8 - last_byte_bits)))
